# Goal of the ETL Lesson

The main goal of this ETL pipelines lesson is to take the [World Bank Project data set](https://datacatalog.worldbank.org/dataset/world-bank-projects-operations) and merge this data with the [World Bank indicator data](https://data.worldbank.org/indicator/SP.POP.TOTL). Then you'll load the merged data into a database.

In the process, you'll need to transform these data sets in different ways. And finally, you'll write a single Python module that reads in these date sets, transforms them, and loads the results into the database all in one step.

# Extracting data from a csv file

The first step in an ETL pipeline is extraction. Data comes in all types of different formats, and you'll practice extracting data from csv files, JSON files, XML files, SQL databases, and the web.

In this first exercise, you'll practice extracting data from a CSV file and then navigating through the results. You'll see that extracting data is not always a straight-forward process.

This exercise contains a series of coding questions for you to solve. If you get stuck, there is a solution file called 1_csv_exercise_solution.ipynb. You can find this solution file by going to File->Open and then clicking on the file name.

# Part 1 projects_data.csv

There are two csv files loaded into this workspace:
* projects_data.csv
* population_data.csv

As a first step, try importing the projects data using the pandas [read_csv method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). The file path is just 'projects_data.csv'. You can see the file if you click on File->Open in the workspace.

In [1]:
# TODO: import the projects_data.csv file using the pandas library
# Store the results in the df_projects variable
import pandas as pd

df_projects = pd.read_csv('projects_data.csv')
df_projects.head()

/Users/meng/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (44) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


id  regionname  ...            Country Unnamed: 56
0  P162228       Other  ...                NaN         NaN
1  P163962      Africa  ...                NaN         NaN
2  P167672  South Asia  ...                NaN         NaN
3  P158768  South Asia  ...                NaN         NaN
4  P161364      Africa  ...  NG;NG;NG;NG;NG;NG         NaN

[5 rows x 57 columns]

Did you get a DType warning? Read about what this warning is in the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.errors.DtypeWarning.html).

Pandas tries to figure out programatically the data type of each column (integer, float, boolean, string). In this case, pandas could not automatically figure out the data type. That is because some columns have multiple data types. In other words, this data is somewhat messy.

You can use the dtype option to specify the data type of each column. Because there are so many columns in this data set, you can set all columns to be strings.

Try reading in the data set again using the read_csv() method. This time, also use the option dtype=str so that pandas treats everything like a string.

In [2]:
# TODO: Read in the projects_data.csv file using the read_csv method 
# and dtype = str option
df_projects = pd.read_csv('projects_data.csv', dtype=str)
df_projects.head()

id  regionname  ...            Country Unnamed: 56
0  P162228       Other  ...                NaN         NaN
1  P163962      Africa  ...                NaN         NaN
2  P167672  South Asia  ...                NaN         NaN
3  P158768  South Asia  ...                NaN         NaN
4  P161364      Africa  ...  NG;NG;NG;NG;NG;NG         NaN

[5 rows x 57 columns]

In [3]:
# TODO: count the number of null values in the data set
# HINT: use the isnull() and sum() methods
df_projects.isnull().sum()

id                              0
regionname                      0
countryname                     0
prodline                        0
lendinginstr                  246
lendinginstrtype              246
envassesmentcategorycode     5811
supplementprojectflg           53
productlinetype                 0
projectstatusdisplay            4
status                          4
project_name                    0
boardapprovaldate            1504
board_approval_month         1504
closingdate                  3349
lendprojectcost               125
ibrdcommamt                     0
idacommamt                      0
totalamt                        0
grantamt                        0
borrower                     5919
impagency                    6097
url                             0
projectdoc                  18248
majorsector_percent         18248
sector1                         0
sector2                      8721
sector3                     11761
sector4                     13872
sector5       

Notice that the number 18248 shows up multiple times. There is also a countryname column with 0 missing values and a Country column with 14045 missing values. This data set has some issues that will need to be solved in the transform part of the pipeline.

Next, output the size of the data frame.

In [4]:
# TODO: output the shape of the data frame
df_projects.shape

(18248, 57)

There are 18248 rows in this data set. Considering many columns had 18248 NaN values, many columns in the data set are filled completely with NaN values.

# Part 2 population_data.csv

Next, use the pandas read_csv method to read in the population_data.csv file. The path to this file is "population_data.csv". When you try to read in this data set using pandas, you'll get an error because there is something wrong with the data.

In [5]:
# TODO: read in the population_data.csv file
# Put the results in a variable called df_population
df_population = pd.read_csv('population_data.csv')
df_population.head()

ParserError: Error tokenizing data. C error: Expected 3 fields in line 5, saw 63


There is something wrong with this data set. You should see an error that says "expected 3 fields in line 5, saw 63". What might have happened? Try printing out the first few lines of the data file to see what the issue might be.

In [7]:
# TODO: Print out the first 10 lines of the data set, line by line.
# HINT: You can't use the read_csv method from pandas
# HINT: to do this manually, you'll need to use pure Python
# HINT: the open(), readline(), and close() methods should be helpful
# HINT: Use a for loop
with open('population_data.csv') as file:
    for _ in range(10):
        print(file.readline())

"Data Source","World Development Indicators",



"Last Updated Date","2018-06-28",



"Country Name","Country Code","Indicator Name","Indicator Code","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017",

"Aruba","ABW","Population, total","SP.POP.TOTL","54211","55438","56225","56695","57032","57360","57715","58055","58386","58726","59063","59440","59840","60243","60528","60657","60586","60366","60103","59980","60096","60567","61345","62201","62836","63026","62644","61833","61079","61032","62149","64622","68235","72504","76700","80324","83200","85451","87277","89005","90853","92898","94992","97017","98737","100031","100832","101220","10135

The first four lines in the file are not properly formatted and don't contain data. Next, read in the data using the read_csv method. But this time, use the skiprows option.

In [8]:
# TODO: read in population data skipping the first four rows
# Put the results in a variable called df_population

df_population = pd.read_csv('population_data.csv', skiprows=4)
df_population.head()

Country Name Country Code  ...        2017 Unnamed: 62
0        Aruba          ABW  ...    105264.0         NaN
1  Afghanistan          AFG  ...  35530081.0         NaN
2       Angola          AGO  ...  29784193.0         NaN
3      Albania          ALB  ...   2873457.0         NaN
4      Andorra          AND  ...     76965.0         NaN

[5 rows x 63 columns]

Make sure to scroll over to see what the last column looks like. That last column, called 'Unnamed: 62', doesn't look very useful and is filled with NaN values.

In [11]:
df_population.shape

(264, 63)

In [9]:
# TODO: Count the number of null values in each column
df_population.isnull().sum()

Country Name        0
Country Code        0
Indicator Name      0
Indicator Code      0
1960                4
                 ... 
2014                2
2015                2
2016                2
2017                2
Unnamed: 62       264
Length: 63, dtype: int64

It looks like every year column has at least one NaN value. 

In [10]:
# TODO: Sum the null values by row
# HINT: In the sum method, use axis=1
df_population.isnull().sum(axis=1)

0      1
1      1
2      1
3      1
4      1
      ..
259    1
260    1
261    1
262    1
263    1
Length: 264, dtype: int64

And it looks like almost every row has only one null value. That is probably from the 'Unnamed: 62' column that doesn't have any relevant information in it. Next, drop the 'Unnamed: 62' column from the data frame.

In [12]:
# TODO: drop the 'Unnamed: 62' column from the data frame, 
# and save the results in the df_population variable

df_population = df_population.drop(columns=['Unnamed: 62'])
df_population.head(1)

Country Name Country Code     Indicator Name  ...      2015      2016      2017
0        Aruba          ABW  Population, total  ...  104341.0  104822.0  105264.0

[1 rows x 62 columns]

In [13]:
# TODO: Now output any row that contains a null value.
# HINT: Use the isnull() and any() methods. You'll want to use axis=1
# in the any method.
df_population.loc[df_population.isnull().any(axis=1)]

Country Name Country Code  ...       2016       2017
67                     Eritrea          ERI  ...        NaN        NaN
108             Not classified          INX  ...        NaN        NaN
125                     Kuwait          KWT  ...  4052584.0  4136528.0
194         West Bank and Gaza          PSE  ...  4551566.0  4684777.0
212                     Serbia          SRB  ...  7058322.0  7022268.0
223  Sint Maarten (Dutch part)          SXM  ...    39969.0    41109.0

[6 rows x 62 columns]

# Conclusion 

This population data doesn't look too bad. Only six rows have missing values. In the transformation part of the lesson, you'll have to deal with these missing values somehow.

If you would like to see the solution file for this exercise, go to File->Open and click on 1_csv_exercise_solution.ipynb.

In the next exercise, you'll practice extracting data json and xml files.